src/providers/dart_rename_provider.ts

```diff
@@ -0,0 +1,91 @@
+"use strict";
+
+import { window, workspace, RenameProvider, WorkspaceEdit, TextDocument, Position, CancellationToken, Uri, TextEdit, Range } from "vscode";
+import { Analyzer } from "../analysis/analyzer";
+import * as as from "../analysis/analysis_server_types";
+
+export class DartRenameProvider implements RenameProvider {
+	private analyzer: Analyzer;
+	constructor(analyzer: Analyzer) {
+		this.analyzer = analyzer;
+	}
+
+	provideRenameEdits(document: TextDocument, position: Position, newName: string, token: CancellationToken): Thenable<WorkspaceEdit> {
+		return this.doRename(document, position, newName, token);
+	}
+
+	private doRename(document: TextDocument, position: Position, newName: string, token: CancellationToken): Thenable<WorkspaceEdit> {
+		return new Promise<WorkspaceEdit>((resolve, reject) => {
+			var wordRange = document.getWordRangeAtPosition(position);
```

Comment:
Docs say for rename:

> It is an error if the range contains anything other than all or part of the name of a single function (including methods, getters and setters), variable (including fields, parameters and local variables), class or function type.

Possibly it's safe to pass 0 or 1 as the length here to avoid relying on `getWordRangeAtPosition` doing a good thing (though maybe if it doesn't, we can improve language options to fix it).